### ISSUE TITLE
Integration tests to understand the ExitCodes 

### ISSUE THREAD

**Issue by yuyiguo** [dmwm/WMArchive#244]
We discussed that we need to run 10-100 integration tests to study the ExitCode. The goal is to understand the exitCode and what is the root exitCode to cause a workflow failed. We will run these tests on testbed and the workflow should have distinct names to be identified.

The problem came out because the exitCodes were interesting to @jenimal could not be found in WMArchive db. She got these error from WMStats(?) or requestManager(?). She was trying to put information from different sources together to identify the cause/location/severity of one error. @ticoann, please comment on how exitCodes were handle in other service. 

@amaltaro you have some tests that can be used for this integration test. Could you please start running them and report the workflow names in this issue.

@vkuznet Addition to the current exitCode plots, should we add a plot that only have the first exitCode for each failed workflow?

**Comment by vkuznet** [comment]
> @vkuznet Addition to the current exitCode plots, should we add a plot that only have the first exitCode for each failed workflow?  

Yuyi, I don't think it's manageable by any means. Think about it. The different
workflows may have different first exit code. What do you really want to plot?
If exitCode value vs time, then exitCodes values are different for different
failed workflows. It total number of exitCodes per given time frame then we'll
need to mix different first exitCodes for different workflows in that time
frame. So your request is not well defined. I'll suggest that you'll think
about this multi-dimensional problem (workflow, site, exitCode, steps vs time)
and understand what is possible to plot in 2D plot.

**Comment by yuyiguo** [comment]
Valentin,
What I suggested was that instead of using all the exitCodes to generate the plots, we could have use only the first exitCode to produce the similar plots, such exitCode vs site or job state. The first exitCode vs time is not really useful. A use case in my mind is that job state per first exitCode, this plot will show what is the main cause for failed jobs. Currently we have the plot with all the exitCode so we cannot use it to identify the root error.

**Comment by vkuznet** [comment]
Yuyi,
again, try to make this plot on a paper, e.g. exitCode vs site.
Let's say you have exitCodes 1-100 and they appear randomly at site A for
different workflows. What do you plot on X and Y axes. The same for
exitCodes vs job state. Please try to make the sketch on a paper
and see if you'll come up with something meaningful on X,Y axes.

And, I said that you can identify the problem, since we have exitCode vs time
plot. Now, you can select a filter, e.g. either workflow or exitCode itself
and see its evolution. The fitler on exitCode itself is what you're looking
for. Choose in left bar the exitCode _metric_. It will plot a exitCode vs time
plot. Then go to top of the page and select specific exitCode value, the
exitCode vs time plot will change to show ONLY this particular exitCode.
If it existed at some time A and then disappeared later you will see it,
and vice versa.

**Comment by ticoann** [comment]
Hi Yuyi,

Here is how wmstats define the exitCode. (Jobs contains several steps - steps are ordered by running stage, i.e first item in the list means it runs first)

1 go through the steps and find first None 0 exit code or 99999 exit code and set that one for exit code for whole job
1. If no other exit code found report 99999 if that exist in some step otherwise report 0.

99999 exit code means  various things (logging failure, frame job report creation failure which can caused by other errors).

For my reference,
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/FwkJobReport/Report.py#L306

For the dashboard we report everything (all the step failure), I am not sure how Dashboard uses that info and display. (Also we report job exit code) but it seem it reports the last step failure.)

For my reference
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMRuntime/DashboardInterface.py#L242

https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMRuntime/DashboardInterface.py#L308

This will call stepEnd and update the exitCode
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/Steps/ExecuteMaster.py#L73

@amaltaro, Alan do you know whether JobExitCode is used in dashboard monitor?

**Comment by ticoann** [comment]
@vkuznet, @yuyiguo, How does WMArchive select exitCode. I was wondering it might be better that we stick the same job exit code as wmstats when we report WMArchive instead of WMArchive having its own logic to select.

**Comment by vkuznet** [comment]
Seangchan,
it is complicated multi-step procedure. I'll outline it here:
- we run myspark script [1] with RecordAggregator.py [2]. This what feeds the
  aggregated.performance MongoDB collection once a day
- then web UI code runs performance method of MongoDB StorageIO layer [3]
  to place individual queries over the data stored in aggregated.performance
  MongoDB
- the results of the previous step are fed into JS library which plot
  specific view, e.g. metric evolution [4]

Now, the selection of exitCodes is done in RecordAggregator.py script here [5]
which basically loop through steps and pick up a first non zero exit code.

It puts it into aggregated.performance document in scope part. You can look-up
any documents by

```
ssh vocms013
source /data/srv/current/apps/wmarchive/etc/profile.d/init.sh
mongo --port 8230
use aggregated
db.performance.findOne()
```

But basically here is an example of scope part of aggregated doc.

```
        "scope" : {
                "task" : "ProductionMergeRAWSIMoutput",
                "jobstate" : "success",
                "end_date" : ISODate("2016-10-13T00:00:00Z"),
                "workflow" : "pdmvserv_TOP-RunIISummer15wmLHEGS-00008_00182_v0__161005_164842_8184",
                "timeframe_precision" : "day",
                "site" : "T2_HU_Budapest",
                "exitStep" : null,
                "host" : "vocms0304.cern.ch",
                "acquisitionEra" : "RunIISummer15wmLHEGS",
                "start_date" : ISODate("2016-10-12T00:00:00Z"),
                "jobtype" : "Merge",
                "exitCode" : null
        },
```

Now, the actual query to MongoDB is done in MongoIO.py performance method, see
[6], which does an aggregated query across all docs in that collection and
group them accordingly. Here is a query:

```
db.performance.aggregate([{$match:{"scope.exitCode":{$ne:null}}}, { $project : {edate: { $dateToString : {format: "%Y-%m-%d", date:"$scope.start_date"} }, ddate: { $dateToString : {format: "%Y-%m-%d", date:"$scope.start_date"} }, dcode:"$scope.exitCode", _id:0} }, {$group: {_id:{'ddate':"$ddate", 'edate':"$edate"}, total:{$sum:1}} }, {$sort: {"_id":1}}, {$project :{'count':"$total", 'average':1, 'label' :{'start_date': '$_id'}}} ])
```

which I decompose for you in plain English:
1. select scope.exitCode which are not null
   `{$match:{"scope.exitCode":{$ne:null}}}`
2. create a new series of docs with ddate, edate and dcode keys
`
{ $project : {edate: { $dateToString : {format: "%Y-%m-%d", date:"$scope.start_date"} }, ddate: { $dateToString : {format: "%Y-%m-%d", date:"$scope.start_date"} }, dcode:"$scope.exitCode", _id:0} }
`
3. group docs according to their ddate/edate stamps and sum up all dcode values as 1
{$group: {_id:{'ddate':"$ddate", 'edate':"$edate"}, total:{$sum:1}} }
`
4. perform sorting on id field which is ddate/edate dict
`
{$sort: {"_id":1}}
`
5. re-map docs into count, average, label fields
`
{$project :{'count':"$total", 'average':1, 'label' :{'start_date': '$_id'}}} ])
`

All these steps produce this type of records:

```
{ "_id" : { "ddate" : "2016-09-01", "edate" : "2016-09-01" }, "count" : 136, "label" : { "start_date" : { "ddate" : "2016-09-01", "edate" : "2016-09-01" } } }
{ "_id" : { "ddate" : "2016-09-02", "edate" : "2016-09-02" }, "count" : 129, "label" : { "start_date" : { "ddate" : "2016-09-02", "edate" : "2016-09-02" } } }
```

These records is what web UI is plotting as exitCode vs time plot.

[1] https://github.com/dmwm/WMArchive/blob/master/bin/myspark
    https://github.com/dmwm/WMArchive/blob/master/src/python/WMArchive/Tools/myspark.py
[2] https://github.com/dmwm/WMArchive/blob/master/src/python/WMArchive/PySpark/RecordAggregator.py
[3] https://github.com/dmwm/WMArchive/blob/master/src/python/WMArchive/Storage/MongoIO.py#L193
[4] https://github.com/dmwm/WMArchive/blob/master/src/js/views/performance.js#L110
[5] https://github.com/dmwm/WMArchive/blob/master/src/python/WMArchive/PySpark/RecordAggregator.py#L44
[6] https://github.com/dmwm/WMArchive/blob/master/src/python/WMArchive/Storage/MongoIO.py#L309

**Comment by yuyiguo** [comment]
"Now, the selection of exitCodes is done in RecordAggregator.py script here [5]
which basically loop through steps and pick up a first non zero exit code."

If I understood correctly, above comment means all the exitCode used in WMArchive ARE the first exitCode a workflow has. BTW, Valentin, that was what I meant use the first ExitCode for the plots.

Can we say WMstats and WMArchive both use the first exitCode that a workflow has?
I think the answer is yes if the order of the exitCodes are arranged by first coming first written in the doc. I did not see there was a reason the doc was not in this order.

This did not explain why Jen saw different exitCodes in these two systems.

**Comment by ticoann** [comment]
> This did not explain why Jen saw different exitCodes in these two systems.

There is one difference, if exitCode is 99999 it will look for next None 0 exitCode. but I am not sure whether that cause the difference.

**Comment by yuyiguo** [comment]
Looked at the WMArchive plot, 99999 was the major exitCode.  We need to change RecordAggregator.py script to search if there is a next no zero exitcode before recording the code as 99999.

**Comment by vkuznet** [comment]
I think it would be a wrong approach to do, since 99999 overwrites existing
exit code of whole job.

The problem is here (from Seangchan explanation):

*
1 go through the steps and find first None 0 exit code or 99999 exit code and set that one for exit code for
whole job
*

we SHOULD NEVER set/overwrite existing exit codes, instead if you want to know if job failed
you check for ANY non-zero exit code in ANY step and declare job to be failed.

Preserving all exit codes will be beneficial for what Jen and others are trying to do,
since we can setup appropriate filter to look-up any exit code.

In my opinion introduction of 99999 defeat the purpose of exit codes completely.
We already have a jobstatus field for that, which we must read as if job contains
any non-zero exit code then entire job is failed.

If I'll change logic of RecordAggregator to look for next non zero exit code then
we still may miss some exit codes, since job overwrites first exit code by 99999.

**Comment by ticoann** [comment]
Hi Valentin, There seems to be misunderstanding on my comment.
Sorry for the confusion.

> *
> 1 go through the steps and find first None 0 exit code or 99999 exit code and set that one for exit code for
> whole job
> *

What I meant is that 99999 exit code is secondary exit code, it doesn't overwrite the current exit code . When there is problem reading or writing the frwjr it uses this exitCode create the new frjw for the next step. Basically for each step if we don't know why job failed (If there is no predefined category for failure, we add 99999 exit code)

i.e. CMSSW finished none 0 exit code (This is system exit code, not the CMS defined exit code) but no job report was made (cmssw job report), then we create the job report with 99999 exit code for that specific step.

So basically, for wmstats exitCode reporting,
1. report first none 0 exit code by traversing each step in order
2. If the 99999 exit code is found in the first step, traverse step to find none 0, none 99999 exit code. If there none 0 (or none 99999) exit code is found, report 99999 exit code. (so we can report more meaning fun exit code even though that is not the fist step job failed.

Hope this is a lilt more clear.

Thanks,

**Comment by vkuznet** [comment]
Seangchan,
I have two statements to make:

On  0, ticoann notifications@github.com wrote:

> Hi Valentin, There seems to be misunderstanding on my comment.
> Sorry for the confusion.
> 
> > *
> > 1 go through the steps and find first None 0 exit code or 99999 exit code and set that one for exit code for
> > whole job
> > *
> 
> What I meant is that 99999 exit code is secondary exit code, it doesn't overwrite the current exit code . When there is problem reading or writing the frwjr it uses this exitCode create the new frjw for the next step. Basically for each step if we don't know why job failed (If there is no predefined category for failure, we add 99999 exit code)

if it is secondary exit code it means that it should be treated equally as any
other exit code, I don't understand the logic of finding non-zero, non-99999
exit code then

> i.e. CMSSW finished none 0 exit code (This is system exit code, not the CMS defined exit code) but no job report was made (cmssw job report), then we create the job report with 99999 exit code for that specific step.

here the root of the problem was CMSSW crash. For that you created 99999 exit
code and assigned it in FWJR. Why do we need to find non-zero/non-99999 code?

> So basically, for wmstats exitCode reporting,
> 1. report first none 0 exit code by traversing each step in order
> 2. If the 99999 exit code is found in the first step, traverse step to find none 0, none 99999 exit code. If there none 0 (or none 99999) exit code is found, report 99999 exit code. (so we can report more meaning fun exit code even though that is not the fist step job failed.

This logic ONLY make sense if step chain order is persistent and represent
reverse chain of steps made in FWJR documents.

My question here is:
Do you preserve the order of steps in step chain? In other words, do all
FWJR documents have the same steps in the same order? Let's say that we
perform 3 steps, A,B,C. What I'm asking is the order of steps:[C,B,A] in
all docs. If this is the case then your logic will make sense, otherwise
it does not make sense to me.

The second question I have is why do you still assign 99999 exit code in your
step #2:

If the 99999 exit code is found in the first step, traverse step to find none 0, none 99999 exit code. If there none 0 (or none 99999) exit code is found, report 99999 exit code. (so we can report more meaning fun exit code even though that is not the fist step job failed.

I don't understand the logic. If 99999 is found in first step it means that the
first step is a root of the problem. Why are you masking it by adding yet
another 99999 code. If order of steps is persistent then you have two
possibility:
- if there is one 99999 (first) exit code you'll report it back.
- if there is another exit code you'll find it and report it back.

Please re-phrase your logic that it would be clear what you're doing with exit
codes and answer the questions I asked about consistency of steps and their
order.

Best,
Valentin.

**Comment by amaltaro** [comment]
I'm not sure I followed all these discussions, but the following questions should shed some light on it:
1. do we always retrieve the steps in the same order? like [cmsRun1, stageOut1, logArch1] or [cmsRun1, cmsRun2, stageout, logarchive]? Seems to be the case looking at https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L286
2. is it possible that the very first exitCode of a step would be 999999? Including cmsRunX steps?

**Comment by yuyiguo** [comment]
@amaltaro will do the "failure test" this week to generate different exitCode for studying the case.

**Comment by amaltaro** [comment]
For the record, this workflow
amaltaro_ReDigi_PUFlag_Integ_Test_WMArchive_161116_213309_6450

should fail all the jobs, though I still have no clue on which exit code that will be.
I removed a `import json` from the SetupCMSSWPset.py runtime script.

**Comment by amaltaro** [comment]
besides 1 LogCollect, all the StepOneProc jobs failed with:
`SCRAMScriptFailure (Exit Code: 50513)`

though I forgot to set the wmarchive to cmsweb instead of testbed... letting another workflow over the night.

**Comment by amaltaro** [comment]
Yuyi, Valentin, a total of 52 docs should have been uploaded to cmsweb.cern.ch/wmarchive for this request name `amaltaro_ReDigi_PUFlag_Integ_Test_WMArchive_161116_232402_6116`.

Where:
- 51 Processing jobs failed with exit code 50513 at T1_US_FNAL_Disk, for cmsRun1 step
- 1 LogCollect job succeeded at T1_US_FNAL_Disk

These failed jobs will also report the following errors:
- exit code 99999 for cmsRun2, logArch1 and stageOut1 steps

Is there anything else needed from our side? What do we (you) do now?

**Comment by vkuznet** [comment]
Alan,
now we need to wait that this data will appear on HDFS, happens tonight, then
wait for aggregated job to pick it up (again tonight, but probably will be
another night) and finally on week-end you'll be able to look-up this
data on wmarchive web ui. We should demonstrate or not that we see
your exit code on exitCode evolution plot.
Valentin.

**Comment by vkuznet** [comment]
Alan,
I run spark job and I don't see this task name between 20161115-20161128 dates. It seems to me that data were not injected to WMArchive properly. Can you create another workflow, inject it to cmsweb.cern.ch/wmarchive, verify on agent side that data was sent to wmarchive and post all details here. Thanks, Valentin.

**Comment by amaltaro** [comment]
Sure. Question though, since we moved to cmsweb as a proxy, any agent can upload data to WMArchive, right? That includes agents that are not at CERN (FNAL agents).

**Comment by vkuznet** [comment]
Yes, since it is under cmsweb any agent/user can upload to wmarchive. The only
requirement would be valid DN.

**Comment by amaltaro** [comment]
Ok, I ran a couple of workflows from cmssrv214.fnal.gov. Both uploaded data into cmsweb.cern.ch/wmarchive

amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161130_192654_9283
had 104 failed jobs with exit code 50513

amaltaro_MonteCarlo_eff_HG1612_WMArchive_161130_192738_7427
had 28 success jobs

We should be able to lookup this data tomorrow in wmarchive.

Just in case, these are the timestamp from the wmagent component (FNAL timezone):
```
2016-11-30 13:15:53,108:140383909619456:INFO:ArchiveDataPoller:Found 98 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:15:56,191:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 98 docs
2016-11-30 13:20:56,334:140383909619456:INFO:ArchiveDataPoller:Found 4 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:20:57,221:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 4 docs
2016-11-30 13:25:57,359:140383909619456:INFO:ArchiveDataPoller:Found 9 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:25:58,551:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 9 docs
2016-11-30 13:30:58,666:140383909619456:INFO:ArchiveDataPoller:Found 8 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:30:59,613:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 8 docs
2016-11-30 13:35:59,676:140383909619456:INFO:ArchiveDataPoller:Found 2 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:36:00,453:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 2 docs
2016-11-30 13:41:00,543:140383909619456:INFO:ArchiveDataPoller:Found 1 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:41:01,294:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 1 docs
2016-11-30 13:46:01,438:140383909619456:INFO:ArchiveDataPoller:Found 2 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:46:02,210:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 2 docs
2016-11-30 13:51:02,269:140383909619456:INFO:ArchiveDataPoller:Found 0 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:56:02,409:140383909619456:INFO:ArchiveDataPoller:Found 3 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 13:56:03,292:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 3 docs
2016-11-30 14:01:03,449:140383909619456:INFO:ArchiveDataPoller:Found 1 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 14:01:04,190:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 1 docs
2016-11-30 14:06:04,249:140383909619456:INFO:ArchiveDataPoller:Found 0 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 14:11:04,295:140383909619456:INFO:ArchiveDataPoller:Found 2 not archived documents from FWRJ db to upload to WMArchive.
2016-11-30 14:11:05,048:140383909619456:INFO:ArchiveDataPoller:...successfully uploaded 2 docs
```

**Comment by vkuznet** [comment]
Alan,
I think you posted not complete information. I can't find your workflow in MongoDB with task name you posted, e.g. amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161130_192654_9283

Instead, I look up one doc in MongoDB and see that task name have two slashes, e.g.

"task" : "/pdmvserv_BPH-RunIISummer16DR80-0019_00037_v0__161108_033028_9027/StepOneProc"

so, I wonder if your posted task name is actually the one used in FWJR? May be that explains why I can't find your original one you posted on Nov 16th.

**Comment by amaltaro** [comment]
I didn't give you the task name, I gave you the workflow name. Tasks have slashes, workflow names don't.

**Comment by vkuznet** [comment]
Alan,
we didn't store workflow name in FWJR, see [1],
but we do store task name. Therefore in order to look-up data I need to know
task name, instead of workflow name or any other info which appears
in [1].

[1] https://github.com/dmwm/WMArchive/blob/master/src/python/WMArchive/Schemas/FWJRProduction.py

**Comment by amaltaro** [comment]
Ok, this is the taskname `/amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161130_192654_9283/DIGI`

However, since we have the `workflow` filter in the web UI, does it mean that data aggregation is somehow broken?

**Comment by vkuznet** [comment]
> Ok, this is the taskname `/amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161130_192654_9283/DIGI`
good, now I can see 104 docs in mongoDB

> However, since we have the `workflow` filter in the web UI, does it mean that data aggregation is somehow broken?

one step at a time Alan.

**Comment by vkuznet** [comment]
> However, since we have the `workflow` filter in the web UI, does it mean that data aggregation is somehow broken?

on a second thought, I think Nils did a deduction from task to workflow name.
It just split split on slash and taking first part.

**Comment by vkuznet** [comment]
Alan, now we can see this workflow on web UI, but there are two problems:
- web UI reports 2x more failed jobs, but this is I think due to the fact that I debug aggregator script and run it multiple times which possibly cause 2x injection. To double check that I suggest you'll generate yet another workflow with small failure.
- all docs show 99999 exitCode in web UI instead of 50513. I fixed the problem in https://github.com/dmwm/WMArchive/pull/265 and to test it I need another set of failed docs.

So, could you please generate yet another small set of failed docs and post their details here. I'll need to know number of docs and task name.

**Comment by amaltaro** [comment]
Sure, this *workflow name* amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161205_111142_2446 should start uploading failed docs to WMArchive within an hour or so. The *taskname* is:
`/amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161205_111142_2446/DIGI`

I also made a couple of comments in your exitcode fix/logic.

**Comment by vkuznet** [comment]
I verified that now we can see failed workflows correctly and search them by error code. It would be nice if Alan @amaltaro will confirm that as well and close the ticket.

**Comment by amaltaro** [comment]
Valentin, testbed is basically unavailable at the moment, but these are the comments I can make at the moment for `amaltaro_StepChain_ReDigi3_HG1612_WMArchive_161205_111142_2446`.
*) the exit code is correct, `50513`
*) job type looks correct, `Processing`
*) I see 416 docs instead of 104 (problem!)
*) the step with exit code `50513` was - I'm almost sure - `cmsRun2`, not `cmsRun1` (problem!)

**Comment by vkuznet** [comment]
Alan,
thanks for report. I look-up all your records on HDFS, so we do have 104
records (not a problem), but it seems that in wmarchive aggregated collection we have
double counting somewhere. It may be because I debugged code and re-run
aggregation. I'll check that.

But I did check that your records had correct step, the cmsRun1 with exit code
`50513` and not cmsRun2. Here is a printout of errors from a random record:

```
>>> for step in rec['steps']: print(step['name'], step.get('errors', ''))
...
(u'stageOut1', [{u'type': u'ReportManipulatingError', u'details': u'Could not find report file for step stageOut1!', u'exitCode': 99999}])
(u'logArch1', [])
(u'cmsRun1', [{u'type': u'SCRAMScriptFailure', u'details': u'SCRAM scripts failed to run!\nPreScriptScramFailure\nMessage: Error running command\n/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.6/bin/python2 -m WMCore.WMRuntime.ScriptInvoke WMTaskSpace.cmsRun1 SetupCMSSWPset \n\n5\n \n        Scram Command Diagnostic:\n        Command : scramv1\n        Architecture: slc6_amd64_gcc481\n        Executed: /cvmfs/cms.cern.ch/slc6_amd64_gcc481/cms/cmssw-patch/CMSSW_7_2_0_patch1/external/slc6_amd64_gcc481/bin/python2.7 -m WMCore.WMRuntime.ScriptInvoke WMTaskSpace.cmsRun1 SetupCMSSWPset \n\n        Exit Status: 5\n        Stdout: None\n        Stderr: None \n\tModuleName : None\n\tMethodName : __init__\n\tClassInstance : None\n\tFileName : /data/srv/wmagent/v1.0.21.patch3/sw/slc6_amd64_gcc493/cms/wmagent/1.0.21.patch3/lib/python2.7/site-packages/WMCore/WMSpec/Steps/WMExecutionFailure.py\n\tClassName : None\n\tLineNumber : 18\n\tErrorNr : 50513\n\nTraceback: \n\n\n Adding last 25 lines of SCRAM error log:\n-------------------------------------------\n# Conditions read from  CMS_CONDITIONS  via FrontierProd \n   Retrieve L1 trigger menu only from XML file \n        L1Menu_Collisions2015_25ns_v1_L1T_Scales_20101224_Imp0_0x102f.xml\n       \nRunning on site \'T1_US_FNAL_Disk\', local PNN: \'T1_US_FNAL_Disk\'\nPileup JSON configuration file: \'/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMTaskSpace/cmsRun1/pileupconf.json\'\nTraceback (most recent call last):\n  File "/cvmfs/cms.cern.ch/slc6_amd64_gcc481/external/python/2.7.6/lib/python2.7/runpy.py", line 162, in _run_module_as_main\n    "__main__", fname, loader, pkg_name)\n  File "/cvmfs/cms.cern.ch/slc6_amd64_gcc481/external/python/2.7.6/lib/python2.7/runpy.py", line 72, in _run_code\n    exec code in run_globals\n  File "/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMCore.zip/WMCore/WMRuntime/ScriptInvoke.py", line 117, in <module>\nRuntimeError: Error invoking script for step WMTaskSpace.cmsRun1\nwithe Script module: SetupCMSSWPset\nglobal name \'json\' is not definedDetails:\n  File "/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMCore.zip/WMCore/WMRuntime/ScriptInvoke.py", line 109, in <module>\n  File "/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMCore.zip/WMCore/WMRuntime/ScriptInvoke.py", line 77, in invoke\n  File "/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMCore.zip/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py", line 693, in __call__\n    self.handlePileup()\n  File "/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMCore.zip/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py", line 379, in handlePileup\n    pileupDict = self._getPileupConfigFromJson()\n  File "/storage/local/data1/condor/execute/dir_2877686/glide_5lg1XY/execute/dir_3964634/job/WMCore.zip/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py", line 493, in _getPileupConfigFromJson\n    pileupDict = json.load(jdata)\n\n', u'exitCode': 50513}])
(u'cmsRun3', [{u'type': u'ReportManipulatingError', u'details': u'Could not find report file for step cmsRun3!', u'exitCode': 99999}])
(u'cmsRun2', [{u'type': u'ReportManipulatingError', u'details': u'Could not find report file for step cmsRun2!', u'exitCode': 99999}])
```

as you can see the code 50513 happened in cmsRun1 step. So I'd say it is not a problem, recharge your memory :)

I'll debug now the problem with double counting. For that I'll wipe out records
in this particular date and re-insert them again via aggregation script.

Thanks,
Valentin.

**Comment by amaltaro** [comment]
Now that we have testbed back, yes, I was wrong. the exit code 50513 happened in cmsRun1.
So the only problem is the amount of docs aggregated. Thanks!

**Comment by vkuznet** [comment]
Alan, try again to look-up your workflow. I fixed double counting issue in a code and regenerated ONLY one day in aggregated db which contains your workflow. Now I see 104 jobs. But I want that you double check that as well.

**Comment by amaltaro** [comment]
Thanks, Valentin. I just had a look and it looks good now. I x-checked those 4 points mentioned above and it looks correct.

**Comment by vkuznet** [comment]
Test were done and implemented in https://github.com/dmwm/WMArchive/tree/master/test/integration closing this issue
